### ISSUE TITLE
Memory leak in .torch.nn.functional.scaled_dot_product_attention

### ISSUE THREAD

**Issue by loua19** [pytorch/pytorch#98940]
### 🐛 Describe the bug

There is a memory leak which occurs when values of dropout above 0.0. When I change this quantity in my code (and only this quantity), memory consumption doubles and cuda training performance reduces by 30%. 

Should be reproducible with any code which uses F.scaled_dot_product_attention

### Versions

PyTorch version: 2.0.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Debian GNU/Linux 10 (buster) (x86_64)
GCC version: (Debian 8.3.0-6) 8.3.0
Clang version: Could not collect
CMake version: version 3.26.3
Libc version: glibc-2.28

Python version: 3.10.8 | packaged by conda-forge | (main, Nov 22 2022, 08:26:04) [GCC 10.4.0] (64-bit runtime)
Python platform: Linux-4.19.0-23-cloud-amd64-x86_64-with-glibc2.28
Is CUDA available: True
CUDA runtime version: 11.3.109
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla V100-SXM2-16GB
Nvidia driver version: 510.47.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
Address sizes:       46 bits physical, 48 bits virtual
CPU(s):              4
On-line CPU(s) list: 0-3
Thread(s) per core:  2
Core(s) per socket:  2
Socket(s):           1
NUMA node(s):        1
Vendor ID:           GenuineIntel
CPU family:          6
Model:               85
Model name:          Intel(R) Xeon(R) CPU @ 2.00GHz
Stepping:            3
CPU MHz:             2000.176
BogoMIPS:            4000.35
Hypervisor vendor:   KVM
Virtualization type: full
L1d cache:           32K
L1i cache:           32K
L2 cache:            1024K
L3 cache:            39424K
NUMA node0 CPU(s):   0-3
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single pti ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities

Versions of relevant libraries:
[pip3] flake8==6.0.0
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.2
[pip3] pytorch-lightning==2.0.1.post0
[pip3] torch==2.0.0
[pip3] torchmetrics==0.11.4
[pip3] triton==2.0.0
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-lightning         2.0.1.post0              pypi_0    pypi
[conda] torch                     2.0.0                    pypi_0    pypi
[conda] torchmetrics              0.11.4                   pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

cc @ngimel @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by ngimel** [comment]
Why do you think it's a memory leak? Does memory usage keep growing?

**Comment by loua19** [comment]
Maybe memory leak was the wrong term. There is definitely an issue with how scaled_dot_product_attention handles dropout values above 0.0. If working correctly I would expect it to slightly reduce gpu memory usage, not double it.

**Comment by ngimel** [comment]
Why would you expect reduced memory usage? Depending on the implementation memory can be larger, and performance can be significantly worse.

**Comment by cpuhrsch** [comment]
@loua19 - Please see [the documentation](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention.html?highlight=scaled_dot_product#torch.nn.functional.scaled_dot_product_attention) on how to use the context manager to figure out whether you're using a fused kernel or not. When you're using a fused kernel the memory footprint and performance is much better. Dropout is part of those conditions.

**Comment by cpuhrsch** [comment]
cc @drisspg

**Comment by loua19** [comment]
@cpuhrsch - Thank you for this update. I thought because dropout support is mentioned in the Flash Attention paper, that this would not be the issue. I understand now that this is probably what is happing.

**Comment by drisspg** [comment]
Dropout is supported by flash_attention but like Christian mentioned the context manager can be used to ensure you are using a fused kernel and will warn with the reasons why a kernel can't be run if for the given inputs to sdpa the choice specified is not possible. Since you think this is what is happening I am going to close this issue.
